#데이터 핸들링 - 판다스

In [ ]:
import pandas as pd

##### 넘파이의 데이터 핸들링은 편하다고 말하기 어렵다. 넘파이는 저수준 API가 대부분이다. 판다스는 많은 부분이 넘파이 기반으로 작성됐는데, 넘파이보다 훨씬 유연하고 편리하게 데이터 핸들링을 가능하게 해준다. 판다스는 파이썬의 리스트, 컬렉션, 넘파이 등의 내부 데이터뿐만 아니라 CSV 등의 파일을 쉽게 DataFrame으로 변경해 데이터의 가공/분석을 편리하게 수행할수 있게 만들어준다.



#####판다스의 핵심 객체는 DataFrame이다. DataFrame은 여러 개의 행과 열로 이뤄진 2차원 데이터를 담는 데이터 **구조체이다**.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
titanic_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/titanic_train.csv')

######read_csv()는 별다른 파라미터 지정이 없으면 파일의 맨 처음 로우를 칼럼명으로 인지하고 칼럼으로 변환한다. 그리고 콤마로 분리된 데이터값들이 해당 칼럼에 맞게 할당됐다. 그런데 맨 왼쪽을 보면 파일에 기재돼 있지 않는 데이터값이 로우 순으로 0, 1, 2, 3……과 같이 순차적으로 표시돼 있음을 알 수 있다. 칼럼명도 표시되지 않는 이 데이터들은 바로 판다스의 Index 객체 값이다. 모든 DataFrame 내의 데이터는 생성되는 순간 고유의 Index 값을 가지게 된다.

In [ ]:
print('titanic 변수 type:', type(titanic_df))
titanic_df

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [ ]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [ ]:
print('DataFrame 크기:', titanic_df.shape)

DataFrame 크기: (891, 12)


In [ ]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


#####describe()은 해당 숫자 칼럼이 숫자형 카테고리 칼럼인지를 판단할 수 있게 도와준다. 카테고리 칼럼은 특정 범주에 속하는 값을 코드화한 칼럼이다. Survived의 경우 min 0, 25〜75%도 0, max도 1 이므로 0과 1로 이뤄진 숫자형 카테고리 칼럼일 것이다. Pclass의 경우도 min이 1, 25%〜75%가2와3, max가 3이므로 1, 2,3으로 이뤄진 숫자형 카테고리 칼럼일 것이다.

In [ ]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

#####value_counts()는 칼럼 값별 데이터 건수를 반환하는 함수로, 데이터의 분포도를 확인하는 데 매우 유용하다.


In [ ]:
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)

Pclass
3    491
1    216
2    184
Name: count, dtype: int64


#####DataFrame의 [  ] 연산자 내부에 칼럼명을 입력하면 해당 칼럼에 해당하는 Series 객체를 반환한다

In [ ]:
titanic_pclass = titanic_df['Pclass']
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


In [ ]:
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

#####이 반환된 Series 객체의 값을 보면 맨 왼쪽이 인덱스값이며, 오른쪽이 데이터값이다. 그런데 이번에는 인덱스가 단순히 0부터 시작하는 순차 값이 아니다. Pclass 칼럼 값이 3, 1, 2를 나타내고 있다. 이처럼 인덱스는 단순히 순차 값과 같은 의미 없는 식별자만 할당하는 것이 아니라 고유성이 보장된다면 의미 있는 데이터값 할당도 가능하다.

In [ ]:
value_counts = titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
Pclass
3    491
1    216
2    184
Name: count, dtype: int64


#####value_counts() 메서드를 사용할 때는 Null 값을 무시하고 결괏값을 내놓기 쉽다는 점을 유의해야 한다. value_counts()는 Null 값을 포함하여 개별 데이터 값의 건수를 계산할지를 dropna 인자로 판단한다. dropna의 기본값은 True이며 Null 값을 무시하고 개별 데이터 값의 건수를 계산한다.
#####Embarked 칼럼은 전체 891건 중에 2건의 데이터가 Null이다. 전체를 합하면 889(644十168+77)가 되기에 2건의 Null 데이터를 제외하고 연산을 수행하였음을 알 수 있다.

In [ ]:
print('titanic_df 데이터 건수:', titanic_df.shape[0])
print('기본 설정인 dropna=True로 value_counts()')
# value_counts()는 디폴트로 dropna=True이므로 value_counts(dropna=True)와 동일
print(titanic_df['Embarked'].value_counts())
print(titanic_df['Embarked'].value_counts(dropna=False))

titanic_df 데이터 건수: 891
기본 설정인 dropna=True로 value_counts()
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


##DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환



#####기본적으로 DataFrame은 파이썬의 리스트, 딕셔너리 그리고 넘파이 ndarray 등 다양한 데이터로부터 생성될 수 있다. 또한 DataFrame은 반대로 파이썬의 리스트, 딕셔너리 그리고 넘파이 ndarray 등으로 변환될 수 있다. 특히 DataFrame과 넘파이 ndarray 상호 간의 변환은 매우 빈번하게 발생한다.



###넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기

In [ ]:
import numpy as np

col_name1 = ['col1'] # 1차원 데이터이므로 칼럼은 1개만 필요!
list1 = [1, 2, 3]
array1 = np.array(list1)
print('array1 shape:', array1.shape)
# 리스트를 이용해 DataFrame 생성
df_list1 = pd.DataFrame(list1, columns=col_name1)
print('1차원 리스트로 만든 DataFrame:\n', df_list1)
# 넘파이 ndarray를 이용해 DataFrame 생성
df_array1 = pd.DataFrame(array1, columns=col_name1)
print('1차원 ndarray로 만든 DataFrame:\n', df_array1)

array1 shape: (3,)
1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


In [ ]:
# 3개의 칼럼명이 필요함!
col_name2=['col1', 'col2', 'col3']

# 2행x3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환
list2 = [[1,2,3],
         [11,12,13]]
array2 = np.array(list2)
print('array2 shape:', array2.shape)
df_list2 = pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame:\n', df_list2)
df_array2 = pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame:\n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13


In [ ]:
# Key는 문자열 칼럼명으로 매핑, Value는 리스트 형(또는 ndarray) 칼럼 데이터로 매핑
dict = {'col1':[1,11], 'col2':[2,22], 'col3':[3,33]}
df_dict = pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame:\n', df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


###DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기


#####많은 머신러닝 패키지가 기본 데이터 형으로 넘파이 ndarray를 사용한다. 따라서 데이터 핸들링은 DataFrame을 이용하더라도 머신러닝 패키지의 입력 인자 등에 적용하기 위해 다시 넘파이 ndarray로 변환하는 경우가 빈번하게 발생한다. *values를 이용한 ndarray로의 변환*은 매우 많이 사용되므로 반드시 기억할 것!

##### 방금 전 생성한 DataFrame 객체인 df_dict를 ndarray로 변환해보자

In [ ]:
array3 = df_dict.values
print('df_dict.values 타입:',type(array3), 'df_dict.values shape:', array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict.values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


##### DataFrame을 리스트와 딕셔너리로 변환해보자. 리스트로의 변환은 values로 얻은 ndarray에 tolist()를 호출하면 된다. 딕셔너리로의 변환은 DataFrame 객체의 to_dict() 메서드를 호출하는데, 인자로 'list'를 입력하면 딕셔너리의 값이 리스트형으로 반환된다.

In [ ]:
# DataFrame을 리스트로 변환
list3 = df_dict.values.tolist()
print('df_dict.values.tolist() 타입:', type(list3))
print(list3)

# DataFrame을 딕셔너리로 변환
dict3 = df_dict.to_dict('list')
print('\n df_dict.to_dict()타입:', type(dict3))
print(dict3)

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict()타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


##DataFrame의 칼럼 데이터 세트 생성과 수정

In [ ]:
# 새로운 칼럼 추가하기
titanic_df['Age_0']=0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

In [ ]:
# 기존 칼럼 series의 데이터를 이용해 새로운 칼럼 series를 만들기
titanic_df['Age_by_10'] = titanic_df['Age']*10
titanic_df['Family_No'] = titanic_df['SibSp'] + titanic_df['Parch']+1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

In [ ]:
titanic_df['Age_by_10'] = titanic_df['Age_by_10'] + 100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

##DataFrame 데이터 삭제

#####drop() 메서드를 이용한다. drop() 메서드의 가장 중요한 parameter는 labels, axis, inplace이다. axis 0은 로우 방향 축, axis 1은 칼럼 방향 축이다. 기존 칼럼 값을 가공해 새로운 칼럼을 만들고 삭제하는 경우가 많다보니 axis=1로 설정하고 드롭하는 경우가 많을 수밖에 없다.

In [ ]:
titanic_drop_df = titanic_df.drop('Age_0', axis=1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

#####삭제됐다고 생각했던 'Age_0' 칼럼이 여전히 존재한다. 그 이유는 앞의 예제 코드에서 inplace = False로 설정했기 때문이다. inplace = False이면 자기 자신의 DataFrame의 데이터는 삭제하지 않으며, 삭제된 결과 DataFrame을 반환한다.

In [ ]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

#####titanic_df에서 ‘Age_0’, ‘Age_by_10’, ‘Family_No’ 세 개의 칼럼이 모두 삭제됐음을 알 수 있다.
#####한 가지 유의할 점은 drop() 시 inplace=True로 설정하면 반환 값이 None(아무 값도 아님)이 된다. 따라서 다음 코드와 같이 inplace=True로 설정한 채로 반환 값을 다시 자신의 DataFrame 객체로 할당하면 안 된다! 즉 titanic_df = titanic_df.drop(['Age_0', 1 Age by lO', 'Family_No'],axis=l, inplace=True)는 titanic_df 객체 변수를 아예 None으로 만들어 버린다는 것이다.

In [ ]:
drop_result = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)
print('inplace=True로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace=True로 drop 후 반환된 값: None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

##### 이번에는 axis=0을 설정해 index 0,1,2 로우를 삭제해보자.

In [ ]:
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))

titanic_df.drop([0,1,2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


## Index 객체

Index 객체는 식별성 데이터를 1차원 array로 가지고 있다. 또한 ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능하다.

In [ ]:
# 원본 파일 다시 로딩
titanic_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/titanic_train.csv')
# Index 객체 추출
indexes = titanic_df.index
print(indexes)
# Index 객체를 실제 값 array로 변환
print('Index 객체 array값:\n', indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232

In [ ]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


#####하지만 한 번 만들어진 DataFrame 및 series의 Index 객체는 함부로 변경할 수 없다.

In [ ]:
indexes[0] = 5

In [ ]:
series_fair = titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series：', sum(series_fair))
print('Fair Series + 3:\n',(series_fair + 3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series： 28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


#####DataFrame 및 Series에 reset_index() 메서드를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 ‘index’라는 새로운 칼럼 명으로 추가합니다

In [ ]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

##### new_value_counts를 보면, 칼럼이 두개가 되므로 Series가 아닌 DataFrame이 반환된다

In [ ]:
print('### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:', type(value_counts))
new_value_counts = value_counts.reset_index(inplace=False)
print('### After reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
Pclass
3    491
1    216
2    184
Name: count, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### After reset_index ###
   Pclass  count
0       3    491
1       1    216
2       2    184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


##데이터 셀렉션 및 필터링

###DataFrame의 [  ]연산자

#####현재 수준에서는 DataFrame 뒤에 있는 [ ]는 칼럼만 지정할 수 있는 ‘칼럼 지정 연산자’로 이해하는 게 혼돈을 막는 가장 좋은 방법이다.
##### 세번째줄 코드에서 오류가 발생하는데, DataFrame 뒤의 [ ]에는 칼럼명을 지정해야 하는데, 0이 칼럼명이 아니기 때문이다.

In [ ]:
print('단일 칼럼 데이터 추출:\n', titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출:\n', titanic_df[['Survived', 'Pclass']].head(3))
print('[ ]안에 숫자 index는 KeyError 오류 발생:\n', titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: 0

##### 앞에서 DataFrame의 [ ] 내에 숫자 값을 입력할 경우 오류가 발생한다고 했는데, 슬라이싱을 이용하면 정확히 원하는 결과를 반환할 수 있다.

In [ ]:
titanic_df[0:2]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

In [ ]:
titanic_df[titanic_df['Pclass'] == 3].head(3)

PassengerId  Survived  Pclass                      Name     Sex   Age  \
0            1         0       3   Braund, Mr. Owen Harris    male  22.0   
2            3         1       3    Heikkinen, Miss. Laina  female  26.0   
4            5         0       3  Allen, Mr. William Henry    male  35.0   

   SibSp  Parch            Ticket   Fare Cabin Embarked  
0      1      0         A/5 21171  7.250   NaN        S  
2      0      0  STON/O2. 3101282  7.925   NaN        S  
4      0      0            373450  8.050   NaN        S

###DataFrame iloc[ ] 연산자

In [ ]:
data = {'Name': ['Chulmin', 'Eunkyung1','inwoong', 'Soobeom'],
        'Year': [2011, 2016, 2015, 2015],
        'Gender': ['Male', 'Female', 'Male', 'Male']
       }
data_df = pd.DataFrame(data, index = ['one', 'two', 'three', 'four'])
data_df

Name  Year  Gender
one      Chulmin  2011    Male
two    Eunkyung1  2016  Female
three    inwoong  2015    Male
four     Soobeom  2015    Male

In [ ]:
data_df.iloc[0,0]

'Chulmin'

#####*iloc[행 위치 정숫값, 열 위치 정숫값]*과 같이 명확하게 DataFrame의 행 위치와 열 위치를 좌표 정숫값 형태로 입력하여 해당 위치에 있는 데이터를 가져올 수 있다.

In [ ]:
# 아래 코드는 오류를 발생시킨다.
data_df.iloc[0,'Name']

In [ ]:
# 아래 코드는 오류를 발생시킨다.
data_df.iloc['one', 0]

In [ ]:
# 다음과 같은 추출도 가능하다.
data_df.iloc[0:2, 0:3]

Name  Year  Gender
one    Chulmin  2011    Male
two  Eunkyung1  2016  Female

In [ ]:
data_df.iloc[0:2, [0,1]]

Name  Year
one    Chulmin  2011
two  Eunkyung1  2016

In [ ]:
# 전체 DataFrame 반환
data_df.iloc[:, :]

Name  Year  Gender
one      Chulmin  2011    Male
two    Eunkyung1  2016  Female
three    inwoong  2015    Male
four     Soobeom  2015    Male

In [ ]:
print("\n 맨 마지막 칼럼 데이터 [:, -1]\n", data_df.iloc[:, -1])
print("\n 맨 마지막 칼럼을 제외한 모든 데이터 [:, :-1]\n", data_df.iloc[:, :-1])


 맨 마지막 칼럼 데이터 [:, -1]
 one        Male
two      Female
three      Male
four       Male
Name: Gender, dtype: object

 맨 마지막 칼럼을 제외한 모든 데이터 [:, :-1]
             Name  Year
one      Chulmin  2011
two    Eunkyung1  2016
three    inwoong  2015
four     Soobeom  2015


###DataFrame loc[ ] 연산자

#####loc[ ]는 명칭(Label) 기반으로 데이터를 추출합니다. 행 위치에는 DataFrame 인덱스 값을, 그리고 열 위치에는 칼럼명을 입력해서 *loc[인덱스값, 칼럼명*]와 같은 형식으로 데이터를 추출할 수 있다.

In [ ]:
data_df.loc['one','Name']

'Chulmin'

In [ ]:
# 다음 코드는 오류를 발생시킨다.
data_df.loc[0,'Name']

#####주의할 점: 일반적으로 슬라이싱을 ‘시작값:종료 값’과 같이 지정하면 시작 값 〜 종료 값-1까지의 범위를 의미하지만,  loc[ ]에 슬라이싱 기호를 적용하면 종료 값-1이 아니라 종료 값까지 포함하는 것을 의미한다. 이는 명칭 기반 인덱싱의 특성 때문이다.

In [ ]:
print('위치기반 iloc slicing\n', data_df.iloc[0:1, 0], '\n')
print('명칭기반 loc slicing\n', data_df.loc['one':'two', 'Name'])

위치기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭기반 loc slicing
 one      Chulmin
two    Eunkyung1
Name: Name, dtype: object


In [ ]:
# 다음과 같은 데이터도 추출 가능하다
data_df.loc['one':'two', ['Name', 'Year']]

Name  Year
one    Chulmin  2011
two  Eunkyung1  2016

In [ ]:
# iloc[ ]과 달리 loc[ ]는 불린 인덱싱이 가능. Year 칼럼의 값이 2014 이상인 모든 데이터를 불린 인덱싱으로 추출
data_df.loc[data_df.Year >= 2014]

Name  Year  Gender
two    Eunkyung1  2016  Female
three    inwoong  2015    Male
four     Soobeom  2015    Male

##### 🌟 iloc[ ]과 loc[ ] 의 특징과 주의할 점을 정리해보자


1.   개별 또는 여러 칼럼 값 전체를 추출하고자 한다면 iloc[ ]나 loc[ ]를 사용하지 않고 data_df['Name']과 같이 DataFrame['칼럼명']만으로 충부나다. 하지만 행과 열을 함께 사용하여 데이터를 추출해야 한다면 iloc[ ]나 loc[ ]를 사용해야 한다.
2. iloc[ ]과 loc[ ]를 이해하기 위해서는 위치 기반 인덱싱과 명칭 기반 인덱싱의 차이를 먼저 이해해야 한다. DataFrame의 인덱스나 칼럼명으로 데이터에 접근하는 것은 명칭 기반 인덱싱이다. 0부터 시작하는 행, 열의 위치 좌표에만 의존하는 것이 위치 기반 인덱싱이다.
3. iloc[ ]는 위치 기반 인덱싱만 가능하다. 따라서 행과 열 위치 값으로 정수형 값을 지정해 원하는 데이터를 반환한다.
4. loc[ ]는 명칭 기반 인덱싱만 가능하다. 따라서 행 위치에 DataFrame 인덱스가 오며, 열 위치에는 칼럼명을 지정해 원하는 데이터를 반환한다.
5. 명칭 기반 인덱싱에서 슬라이싱을 '시작점:종료점'으로 지정할 때 시작점에서 종료점을 포함한 위치에 있는 데이터를 반환한다.



###불린 인덱싱

#####불린 인덱싱은 매우 편리한 데이터 필터링 방식이다. 오히려 iloc나 loc와 같이 명확히 인덱싱을 지정하는 방식보다는 불린 인덱싱에 의존해 데이터를 가져오는 경우가 더 많다.

In [ ]:
titanic_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/titanic_train.csv')
titanic_boolean = titanic_df[titanic_df['Age']>60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass                                       Name  \
33            34         0       2                      Wheadon, Mr. Edward H   
54            55         0       1             Ostby, Mr. Engelhart Cornelius   
96            97         0       1                  Goldschmidt, Mr. George B   
116          117         0       3                       Connors, Mr. Patrick   
170          171         0       1                  Van der hoef, Mr. Wyckoff   
252          253         0       1                  Stead, Mr. William Thomas   
275          276         1       1          Andrews, Miss. Kornelia Theodosia   
280          281         0       3                           Duane, Mr. Frank   
326          327         0       3                  Nysveen, Mr. Johan Hansen   
438          439         0       1                          Fortune, Mr. Mark   
456          457         0       1                  Millet, Mr. Francis Davis   
483          484         1       3                     Turkula, Mrs. (Hedwig)   
493          494         0       1                    Artagaveytia, Mr. Ramon   
545          546         0       1               Nicholson, Mr. Arthur Ernest   
555          556         0       1                         Wright, Mr. George   
570          571         1       2                         Harris, Mr. George   
625          626         0       1                      Sutton, Mr. Frederick   
630          631         1       1       Barkworth, Mr. Algernon Henry Wilson   
672          673         0       2                Mitchell, Mr. Henry Michael   
745          746         0       1               Crosby, Capt. Edward Gifford   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   
851          852         0       3                        Svensson, Mr. Johan   

        Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked  
33     male  66.0      0      0   C.A. 24579   10.5000          NaN        S  
54     male  65.0      0      1       113509   61.9792          B30        C  
96     male  71.0      0      0     PC 17754   34.6542           A5        C  
116    male  70.5      0      0       370369    7.7500          NaN        Q  
170    male  61.0      0      0       111240   33.5000          B19        S  
252    male  62.0      0      0       113514   26.5500          C87        S  
275  female  63.0      1      0        13502   77.9583           D7        S  
280    male  65.0      0      0       336439    7.7500          NaN        Q  
326    male  61.0      0      0       345364    6.2375          NaN        S  
438    male  64.0      1      4        19950  263.0000  C23 C25 C27        S  
456    male  65.0      0      0        13509   26.5500          E38        S  
483  female  63.0      0      0         4134    9.5875          NaN        S  
493    male  71.0      0      0     PC 17609   49.5042          NaN        C  
545    male  64.0      0      0          693   26.0000          NaN        S  
555    male  62.0      0      0       113807   26.5500          NaN        S  
570    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S  
625    male  61.0      0      0        36963   32.3208          D50        S  
630    male  80.0      0      0        27042   30.0000          A23        S  
672    male  70.0      0      0   C.A. 24580   10.5000          NaN        S  
745    male  70.0      1      1    WE/P 5735   71.0000          B22        S  
829  female  62.0      0      0       113572   80.0000          B28      NaN  
851    male  74.0      0      0       347060    7.7750          NaN        S

In [ ]:
titanic_df[titanic_df['Age']>60][['Name','Age']].head(3)

Name   Age
33           Wheadon, Mr. Edward H  66.0
54  Ostby, Mr. Engelhart Cornelius  65.0
96       Goldschmidt, Mr. George B  71.0

In [ ]:
titanic_df.loc[titanic_df['Age']>60, ['Name','Age']].head(3)

Name   Age
33           Wheadon, Mr. Edward H  66.0
54  Ostby, Mr. Engelhart Cornelius  65.0
96       Goldschmidt, Mr. George B  71.0


1.  and 조건일 때는 &  
2.  or 조건일 |
3.  not 조건일 때는 ~

In [ ]:
titanic_df[(titanic_df['Age']>60) & (titanic_df['Pclass']==1) & (titanic_df['Sex']=='female')]

PassengerId  Survived  Pclass                                       Name  \
275          276         1       1          Andrews, Miss. Kornelia Theodosia   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
275  female  63.0      1      0   13502  77.9583    D7        S  
829  female  62.0      0      0  113572  80.0000   B28      NaN

In [ ]:
cond1 = titanic_df['Age']>60
cond2 = titanic_df['Pclass']==1
cond3 = titanic_df['Sex']=='female'
titanic_df[cond1 & cond2 & cond3]

PassengerId  Survived  Pclass                                       Name  \
275          276         1       1          Andrews, Miss. Kornelia Theodosia   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
275  female  63.0      1      0   13502  77.9583    D7        S  
829  female  62.0      0      0  113572  80.0000   B28      NaN

##정렬, Aggregation 함수, GroupBy 적용

###DataFrame, Series의 정렬 - sort_values()

In [ ]:
titanic_sorted = titanic_df.sort_values(by=['Name'])
titanic_sorted.head(3)

PassengerId  Survived  Pclass                              Name     Sex  \
845          846         0       3               Abbing, Mr. Anthony    male   
746          747         0       3       Abbott, Mr. Rossmore Edward    male   
279          280         1       3  Abbott, Mrs. Stanton (Rosa Hunt)  female   

      Age  SibSp  Parch     Ticket   Fare Cabin Embarked  
845  42.0      0      0  C.A. 5547   7.55   NaN        S  
746  16.0      1      1  C.A. 2673  20.25   NaN        S  
279  35.0      1      1  C.A. 2673  20.25   NaN        S

In [ ]:
# Pclass와 Name를 내림차순으로 정렬한 결과 반환
titanic_sorted = titanic_df.sort_values(by=['Pclass', 'Name'], ascending=False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass                             Name   Sex  \
868          869         0       3      van Melkebeke, Mr. Philemon  male   
153          154         0       3  van Billiard, Mr. Austin Blyler  male   
282          283         0       3        de Pelsmaeker, Mr. Alfons  male   

      Age  SibSp  Parch    Ticket  Fare Cabin Embarked  
868   NaN      0      0    345777   9.5   NaN        S  
153  40.5      0      2  A/5. 851  14.5   NaN        S  
282  16.0      0      0    345778   9.5   NaN        S

###Aggregation 함수 적용

#####단, count()는 Null 값을 반영하지 않은 결과를 반환한다. 때문에 Null 값이 있는 Age, Cabin, Embarked 칼럼은 count()의 결과 값이 다르다

In [ ]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [ ]:
titanic_df[['Age','Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

###groupby() 적용

#####groupby()는 분석 작업에서 매우 많이 사용하는 함수이다. 다음 예제에서 groupby(by='Pclass')를 호출하면 Pclass 칼럼 기준으로 GroupBy된 DataFrameGroupby 객체를 반환한다.

In [ ]:
titanic_groupby = titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [ ]:
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass                                                                      
1               216       216   216  216  186    216    216     216   216   
2               184       184   184  184  173    184    184     184   184   
3               491       491   491  491  355    491    491     491   491   

        Cabin  Embarked  
Pclass                   
1         176       214  
2          16       184  
3          12       491

In [ ]:
titanic_groupby =titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

#####DataFrame groupby()의 경우 적용하려는 여러 개의 aggregation 함수명을 DataFrameGroupBy 객체의 agg( ) 내에 인자로 입력해서 사용한다

In [ ]:
titanic_df.groupby('Pclass')['Age'].agg([max, min])

<ipython-input-50-b719d79acb8d>:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  titanic_df.groupby('Pclass')['Age'].agg([max, min])
<ipython-input-50-b719d79acb8d>:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  titanic_df.groupby('Pclass')['Age'].agg([max, min])


max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

#####여러 개의 칼럼이 서로 다른 aggregation 함수를 groupby에서 호출하고 싶을 때

In [ ]:
agg_format = {'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

##결손 데이터 처리하기

###isna()로 결손 데이터 여부 확인

##### DataFrame에 isna()를 수행하면 모든 칼럼의 값이 NaN인지 아닌지를 True나 False로 알려준다.

In [ ]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0        False     False   False  False  False  False  False  False   False   
1        False     False   False  False  False  False  False  False   False   
2        False     False   False  False  False  False  False  False   False   

    Fare  Cabin  Embarked  
0  False   True     False  
1  False  False     False  
2  False   True     False

In [ ]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

###fillna()로 결손 데이터 대체하기

#####주의해야 할 점은 fillna()를 이용해 반환 값을 다시 받거나 inplace=True 파라미터를 fillna()에 추가해야 실제 데이터 세트 값이 변경된다는 점이다. titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C0OOO')와 같이 fillna() 반환 값을 다시 titanic_df['Cabin']로 정해주거나 반환 변숫값을 지정하지 않으려면 titanic_df['Cabin'].fillna('COOO', inplace=True)와같이 inplace=True를 설정해야 한다.

In [ ]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  C000        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  C000        S

In [ ]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

##apply lambda 식으로 데이터 가공

In [ ]:
def get_square(a):
  return a**2

print('3의 제곱은:', get_square(3))

3의 제곱은: 9


#####lambda는 이러한 함수의 선언과 함수 내의 처리를 한 줄의 식으로 쉽게 변환하는 식이다. ‘:’의 왼쪽에 있는 x는 입력 인자를 가리키며, 오른쪽은 입력 인자의 계산식이다.

In [ ]:
lambda_square = lambda x : x**2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


##### lambda 식을 이용할 때 여러 개의 값을 입력 인자로 사용해야 할 경우에는 보통 map() 함수를 결합해서 사용합니다.


In [ ]:
a = [1,2,3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]

In [ ]:
titanic_df['Name_len'] = titanic_df['Name'].apply(lambda x : len(x))
titanic_df[['Name', 'Name_len']].head(3)

Name  Name_len
0                            Braund, Mr. Owen Harris        23
1  Cumings, Mrs. John Bradley (Florence Briggs Th...        51
2                             Heikkinen, Miss. Laina        22

##### lambda 식은 if else를 지원하는데, 주의할 점이 있다. if절의 경우 if 식보다 반환 값을 먼저 기술해야 한다. else의 경우는 else 식이 먼저 나오고 반환값이 나중에 오면 된다.

In [ ]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else 'Adult')
titanic_df[['Age', 'Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

##### if.else만 지원하고 if,else if, else와 같이 else if는 지원하지 않는다. else if를 이용하기 위해서는 else절을 ()로 내포해 ()내에서 다시 if else 적용해 사용한다.

In [ ]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x<= 15 else ('Adult' if x <= 60 else 'Elderly'))
titanic_df['Age_cat'].value_counts()

Age_cat
Adult      786
Child       83
Elderly     22
Name: count, dtype: int64

##### 더 세분화된 분류를 하는 경우는 별도의 함수를 만드는 것이 더 나을 수 있다.


In [ ]:
# 나이에 따라 더 세분화된 분류를 수행하는 함수 생성
def get_category(age):
  cat=''
  if age <= 5: cat = 'Baby'
  elif age <= 12: cat = 'Child'
  elif age <= 18: cat= 'Teenager'
  elif age <= 25: cat = 'Student'
  elif age <= 35: cat = 'Young Adult'
  elif age <= 60: cat = 'Adult'
  else : cat = 'Elderly'

  return cat

# lambda 식에 위에서 생성한 get_category() 함수를 반환값으로 지정
# get_category(X)는 입력값으로 'Age' 칼럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age','Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult

##### ☑️정리- 넘파이와 판다스는 머신러닝 프로그램을 직접 만들면서 이해하고, 모르는 부분들을 인터넷 검색 등으로 찾아보는 게 학습을 위한 제일 빠른 방법이다. , 실제로 ML 모델을 생성하고 예측을 수행하는 데 있어서 ML 알고리즘이 차지하는 비중보다 데이터를 전처리하고 적절한 피처를 가공/추출하는 부분이 훨씬 많은 비중을 차지하게 된다. 따라서 사이킷런에 대한 이해도 중요하지만, 넘파이, 판다스, 맷플롯립, 시본과 같이 파이썬 머신러닝 생태계를 이루는 다양한 패키지에 대한 이해 역시 중요하다!

#사이킷런 소개와 특징

In [ ]:
pip install scikit-learn==1.0.2

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.5/26.5 MB 63.9 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.3.2
    Uninstalling scikit-learn-1.3.2:
      Successfully uninstalled scikit-learn-1.3.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
bigframes 1.15.0 requires scikit-learn>=1.2.2, but you have scikit-learn 1.0.2 which is incompatible.


In [ ]:
import sklearn

print(sklearn.__version__)

1.0.2
